#################################################
## Configuration file for PhyloCap
#################################################

#Directories and input files
#########################
# *** Full paths should be used whenever possible
#The main working directory
working.directory = "/Path/to/where/the/stuff/is/at"
#The file for the contaminant genomes (Genome, Accession columns); use NULL if download.contaminant.genomes = F
contaminant.genome.list = "decontamination_database.csv"
#The name for the dataset
dataset.name = "Descriptive-Name"

#Raw read locations
#########################
#The file rename (File, Sample columns) for organizing reads or dropbox download. NA if reads already organized
sample.file = "file_rename.csv"
#TRUE = to subset reads from a larger pool using the "sample.file" above
organize.reads = FALSE
#The input raw read directory
read.directory = "/Path/to/where/the/raw/reads/are"
#The name for the processed reads folder
processed.reads = "processed-reads"
#TRUE to save a summary csv file of the raw sequence data
summary.fastq = TRUE
#TRUE to download files from personal dropbox folder using token and read path
dropbox.download = FALSE
#the dropbox directory your files are all contained within if dropbox.download = TRUE
dropbox.directory = "/Dropbox/Path/to/Reads"
#token file previously generated by the rdrop2 package (see this package for help)
dropbox.token = "/Local/Path/to/Token"

#Global settings
#########################
#number of threads
threads = 16
#Amount of memory to allocate in GB
memory = 120
# TRUE to overwrite previous runs. FALSE the script will resume but will not delete anything.
overwrite = FALSE
#Print verbose output for each function
quiet = TRUE

#FASTP read cleaning
#########################
# = TRUE to run all processing steps at once (much faster). Overrides settings below.
# = FALSE to run the separate analyses with TRUE below
fastp.complete = TRUE
#TRUE = to run adaptor removal on reads
remove.adaptors = TRUE
#TRUE to remove exact PCR duplicates
remove.duplicate.reads = TRUE
#TRUE to correct errors using the other read pair
error.correction = TRUE
#Trims low quality ends off of reads
quality.trim.reads = TRUE

#Other read processing tasks
############################
#normalizes read depth to facilitate assembly
normalize.reads = TRUE
#Merge paired end reads, helps with assembly
merge.pe.reads = TRUE

#Decontamination settings
#########################
#Remove contamination
decontamination = TRUE
#Download contaminat genomes from genbank if TRUE
download.contaminant.genomes = TRUE
#Alternatively, a path can be set to a downloaded set of contaminant genomes if downloading does not work; NULL if downloading
decontamination.path = NULL
#include the human genome? Unless human is study organism or UCEs in mammals are used
include.human = TRUE
#Include the univec contaminant database?
include.univec = TRUE
#The matching proportion of bases for a contaminant hit and removal
decontamination.match = 0.99

#Assembly settings
#########################
#Denovo assembly with spades
denovo.assembly = TRUE
#The selected k-mer values for spades
spades.kmer.values = c(33,55,77,99,127)
#Whether to use mismatch corrector (requires a lot of RAM and resources)
spades.mismatch.corrector = TRUE
#Whether to save the corrected reads
save.corrected.reads = FALSE
#Clean up large files *** ADD INTO ANALYSES
clean.up.spades = FALSE

#Program paths
#########################
### *** Modify any of these from NULL to the path that the program is found if R is not detecting system paths
### e.g. fastp.path = "/conda/PhyloCap/bin
### Alternatively, if prereqs installed following tutorial, can set to single conda environment path
conda.env = "/Path/to/conda/env/bin"
fastp.path = conda.env
samtools.path = conda.env
bwa.path = conda.env
spades.path = conda.env
bbnorm.path = conda.env
blast.path = conda.env

